pkg/agent/instancevalidator.go

```diff
@@ -96,22 +96,23 @@ func (j *instanceValidator) getServiceInstanceCount(attName, attValue string) in
 }
 
 func (j *instanceValidator) deleteServiceInstanceOrService(ri *apiV1.ResourceInstance, primaryKey, apiID string) {
-	msg := ""
-	var err error
-	var agentError *utilErrors.AgentError
+	// delete if it is an api service instance
+	log.Infof("API no longer exists on the dataplane, deleting the catalog item %s", ri.Title)
+	msg := "Deleted catalog item %s from Amplify Central"
 
-	defer func() {
-		// remove the api service instance from the cache for both scenarios
-		if j.isAgentPollMode {
-			// In GRPC mode delete is done on receiving delete event from service
-			agent.cacheManager.DeleteAPIServiceInstance(ri.Metadata.ID)
-		}
-	}()
+	err := agent.apicClient.DeleteAPIServiceInstance(ri.Name)
+	if err != nil {
+		log.Error(utilErrors.Wrap(ErrDeletingCatalogItem, err.Error()).FormatError(ri.Title))
+		return
+	}
+	if j.isAgentPollMode {
```

Comment:
Seems it would cause issue in gRPC mode where the validator deletes the instance and the item is not removed from the cache so shouldDeleteService is not going to remove the APIService. When the handler will receive the delete resource event, the instance will be removed from the cache by handler, however on next validator execution the above DeleteAPIServiceInstance call is going to fail since it will not find the instance and the service will not be removed.

May be we can remove the isAgentPollMode check here

Reply:
Removed, will turn on grpc in my env to test this today